pkg/resource/open_id_connect_provider/hooks.go

```diff
@@ -0,0 +1,297 @@
+		msg := fmt.Sprintf("Immutable Spec fields have been modified: %s", strings.Join(immutableFieldChanges, ","))
+		return nil, ackerr.NewTerminalError(fmt.Errorf(msg))
+	}
+
+	if delta.DifferentAt("Spec.ThumbprintList") {
+		// Update the thumbprint list
+		thumbprintInput, err := rm.newUpdateThumbprintRequestPayload(ctx, desired)
+		if err != nil {
+			return nil, err
+		}
+
+		var thumbprintResp *svcsdk.UpdateOpenIDConnectProviderThumbprintOutput
+		_ = thumbprintResp
+		thumbprintResp, err = rm.sdkapi.UpdateOpenIDConnectProviderThumbprintWithContext(ctx, thumbprintInput)
+		rm.metrics.RecordAPICall("UPDATE", "UpdateOpenIDConnectProviderThumbprint", err)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if delta.DifferentAt("Spec.ClientIDList") {
+		// Update the client ID list
+		// here we only have an "add" and a "remove"
+		// https://docs.aws.amazon.com/sdk-for-go/api/service/iam/#IAM.AddClientIDToOpenIDConnectProvider and
+		// https://docs.aws.amazon.com/sdk-for-go/api/service/iam/#IAM.RemoveClientIDFromOpenIDConnectProvider
+		// so we have to compute the diff ourselves
+		desiredClientIDs := map[string]bool{}
+		latestClientIDs := map[string]bool{}
+		for _, desiredClientID := range desired.ko.Spec.ClientIDList {
+			desiredClientIDs[*desiredClientID] = true
+		}
+		for _, latestClientID := range latest.ko.Spec.ClientIDList {
+			latestClientIDs[*latestClientID] = true
+		}
+
+		for desiredClientID, _ := range desiredClientIDs {
+			_, hasLatest := latestClientIDs[desiredClientID]
+			if !hasLatest {
+				// clientID is to be added
+
+				addClientIDInput, err := rm.newAddClientIDRequestPayload(ctx, desired, &desiredClientID)
+				if err != nil {
+					return nil, err
+				}
+
+				var addClientIDResp *svcsdk.AddClientIDToOpenIDConnectProviderOutput
+				_ = addClientIDResp
+				addClientIDResp, err = rm.sdkapi.AddClientIDToOpenIDConnectProviderWithContext(ctx, addClientIDInput)
+				rm.metrics.RecordAPICall("UPDATE", "AddClientIDToOpenIDConnectProvider", err)
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				delete(desiredClientIDs, desiredClientID)
+				delete(latestClientIDs, desiredClientID)
+			}
+		}
+		for latestClientID, _ := range latestClientIDs {
+			// clientID is to be removed
+			removeClientIDInput, err := rm.newRemoveClientIDRequestPayload(ctx, desired, &latestClientID)
+			if err != nil {
+				return nil, err
+			}
+
+			var removeClientIDResp *svcsdk.RemoveClientIDFromOpenIDConnectProviderOutput
+			_ = removeClientIDResp
+			removeClientIDResp, err = rm.sdkapi.RemoveClientIDFromOpenIDConnectProviderWithContext(ctx, removeClientIDInput)
+			rm.metrics.RecordAPICall("UPDATE", "removeClientIDFromOpenIDConnectProvider", err)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Merge in the information we read from the API call above to the copy of
+	// the original Kubernetes object we passed to the function
+	ko := desired.ko.DeepCopy()
+
+	rm.setStatusDefaults(ko)
```

Comment:
I don't think this is needed. Since the update operations don't mess with the desired state object.

Reply:
Removed the deep copy in favor of just using `desired`